### PR TITLE
Improve set formatter dialog

### LINKED ISSUE #1035: Improve set formatter dialog for taurus forms
The `TaurusForm` allows the user to change the formatter (per-item and to all items), but the dialog is very crude (just a text input), which is not user friendly.

It would be nice to improve this dialog with, at least the following:

- add a user-friendly tooltip explaining usage
- make it an editable combobox with pre-set options such as `{:2.3e}`, `{:.5f}`, `TangoFormatter`, ,...

Note: the formatter classes to be offered should be gathered from a list of "registered formatters"  (i.e. an entry point (e.g. `taurus.qt.formatters`) should be declared and plugins should register their formatters into it.

### PR DESCRIPTION
**Pull request by cmft**
* This PR adds a new entry point for registering Taurus formatters.
* Improves the usability of taurus formatter dialog (using the entry points)

Fixes #1035

### CODE REVIEW

File: lib/taurus/qt/qtgui/base/taurusbase.py [hunk 1]
```diff
@@ -66,6 +68,23 @@
...
+for __p in pkg_resources.iter_entry_points('taurus.qt.qtgui.base.formatter'):
+    try:
+        scheme_formatters[__p.name] = '{}.{}'.format(__p.module_name,
+                                                     __p.attrs[0])
+    except Exception as e:
+        warning('Could not load formatter plugin "%s". Reason: %s',
+                __p.module_name, e)
+
+# ------------------------------------------------------------------------
+
```
Comment (cpascual) [inline lib/taurus/qt/qtgui/base/taurusbase.py]: In terms of implementation, I think that this is not the best way of collecting the registered formatters for the following reasons:
- I would call the entry point `taurus.qt.qtgui.base.formatters` (note the final "s"), or even `taurus.qt.formatters`
- The iteration and collection can be simplified (see proposed snippet below)
- The discovery is better delayed until the moment is needed (e.g. define getter for the formatters instead of loading them when importing the module).

In summary, I would do something more or less like this:
```python
def get_registered_formatters():
    formatters = {}
    for ep in pkg_resources.iter_entry_points('taurus.qt.qtgui.base.formatters'):
        try:
            formatters[ep.name] = ep.load()
        except Exception as e:
            taurus.warning('Cannot load "%s" formatter. Reason: %r', ep.name, e)
    return formatters
```

File: lib/taurus/qt/qtgui/base/taurusbase.py [hunk 2]
```diff
@@ -1342,6 +1361,8 @@ def __init__(self, name='', parent=None, designMode=False, **kwargs):
         self.call__init__(TaurusBaseComponent, name,
                           parent=parent, designMode=designMode)
         self._setText = self._findSetTextMethod()
+        self._formaters = scheme_formatters
+        self._formaters['{:2.3e}'] = '{:2.3e}'
```
Comment (cpascual) [inline lib/taurus/qt/qtgui/base/taurusbase.py]: spelling: use "formatter" instead of "formater" (check also other usages)

Comment (cpascual) [inline lib/taurus/qt/qtgui/base/taurusbase.py]: Similar to what I was mentioning before: I would not keep a list of loaded formatters referenced by the TaurusBaseWidget. In other words, instead of keeping `self._formaters` I would just call the getter whenever is needed (e.g. in `showFormatterDlg()`)

Comment (cpascual) [inline lib/taurus/qt/qtgui/base/taurusbase.py]: Instead of defining the exponential formatter here, I would simply register it using the entry point (just as the default formatter)

File: setup.py [hunk 3]
```diff
@@ -118,10 +118,16 @@ def get_release_info():
...
+formatters = [
+    'default = taurus.qt.qtgui.base.taurusbase:defaultFormatter',
+    'tango = taurus.core.tango.util.formatter:tangoFormatter',
+]
+
 entry_points = {
     'console_scripts': console_scripts,
     'taurus.cli.subcommands': taurus_subcommands,
     'taurus.qt.qtgui.panel.TaurusModelSelector.items': model_selectors,
+    'taurus.qt.qtgui.base.formatter': formatters,
```
Comment (cpascual) [inline setup.py]: As mentioned above, use "formatter**s**"

File: lib/taurus/qt/qtgui/base/taurusbase.py [hunk 4]
```diff
@@ -1350,13 +1371,27 @@ def showFormatterDlg(self):
         (in string version) or None
         """
         current_format = self.getFormat()
+        ind = self._formaters.values().index(current_format)
+
+        formatter, ok = Qt.QInputDialog.getItem(self, "Set formatter",
```
Comment (cpascual) [inline lib/taurus/qt/qtgui/base/taurusbase.py]: Since the dialog does not allow to customize the tooltip, use the "label text" to give the hint (maybe using an html link to the formatter docs?)

File: lib/taurus/qt/qtgui/base/taurusbase.py (hunk 1 above)
Comment (cpascual) [reply lib/taurus/qt/qtgui/base/taurusbase.py]: Note the getter could also be implemented as a static method of TaurusBaseComponent (to make it available to any widget

**Comment by cmft** [comment]
Done

File: lib/taurus/qt/qtgui/base/taurusbase.py (hunk 2 above)
Comment (cpascual) [reply lib/taurus/qt/qtgui/base/taurusbase.py]: After discussing with @cmft , we agreed that keeping a list may still be needed in order to offer persistence in the choice of formatter hwen the user sets a custom formatter

File: lib/taurus/qt/qtgui/base/taurusbase.py [hunk 5]
```diff
@@ -1342,21 +1366,52 @@ def __init__(self, name='', parent=None, designMode=False, **kwargs):
...
+            self._formatters = TaurusBaseWidget.get_registered_formatters()
+
+        # Find index of the current formatter
+        try:
+            ind = self._formatters.values().index(self.FORMAT)
+        except ValueError:
+            ind = 0
+
+        uri = "http://taurus-scada.org/devel/api/taurus/qt/qtgui/base/_TaurusBaseComponent.html"
+        help = "<a href=\"{}\"> Find more info in doc</a>".format(uri)
```
Comment (cpascual) [inline lib/taurus/qt/qtgui/base/taurusbase.py]: Not crytical, but the link does not launch a browser in my system when clicked (but it allows to right-click and copy the URL which , when pasted on a browser works as expected)

File: lib/taurus/qt/qtgui/base/taurusbase.py [hunk 6]
```diff
@@ -1342,21 +1366,52 @@ def __init__(self, name='', parent=None, designMode=False, **kwargs):
...
+
+        msg = "Choose/Enter a formatter: {}".format(help)
```
Comment (cpascual) [inline lib/taurus/qt/qtgui/base/taurusbase.py]: I would replace `msg` and `help` by:

```python
msg = 'Choose/Enter a formatter (<a href="{}">help</a>)'.format(uri)
```

File: lib/taurus/qt/qtgui/base/taurusbase.py [hunk 7]
```diff
@@ -1342,21 +1366,52 @@ def __init__(self, name='', parent=None, designMode=False, **kwargs):
...
         (in string version) or None
         """
-        current_format = self.getFormat()
+        if self._formatters is None:
+            # initialize cache
+            self._formatters = TaurusBaseWidget.get_registered_formatters()
+
+        # Find index of the current formatter
+        try:
+            ind = self._formatters.values().index(self.FORMAT)
```
Comment (cpascual) [inline lib/taurus/qt/qtgui/base/taurusbase.py]: This fails with py3

Comment (cpascual) [reply lib/taurus/qt/qtgui/base/taurusbase.py]: fixed in 5e32a69

File: lib/taurus/qt/qtgui/base/taurusbase.py (hunk 6 above)
Comment (cpascual) [reply lib/taurus/qt/qtgui/base/taurusbase.py]: fixed in 5e32a69

**Comment by cpascual** [comment]
I did some heavy refactoring. Can some @taurus-org/integrators  test it?

Here is a snippet of a code that can be used to test:

```python
from taurus.qt.qtgui.application import TaurusApplication
from taurus.qt.qtgui.panel import TaurusForm
import taurus

def myformatter(*a, **kw):
    return "**{1.2e}**"

if __name__ == '__main__':
    import sys
    app = TaurusApplication(cmd_line_parser=None)
    w = TaurusForm()
    w.setModifiableByUser(True)
    model = ['sys/tg_test/1/float_scalar', 'sys/tg_test/1/double_scalar']
    w.setModel(model)
    w.show()

    from taurus.core.tango.util import tangoFormatter
    # w[0].setFormat(tangoFormatter)
    # w[0].setFormat('taurus.core.tango.util.tangoFormatter')
    # w[0].setFormat(myformatter)
    w[0].setFormat('///{:1.2f}///')

    # show setFormatter dialog
    w[0].onSetFormatter()
    sys.exit(app.exec_())
```

I just see one thing missing: if the selected formatter happens to be `None` (which is often the case for the class formatter), `onSetFormatter` does not set it. This requires either deprecating the `onSetFormatter`/`showFormatterDlg` API or breaking it. I do not expect anyone having reimplemented these methods in their own code, so I vote for breaking in this corner case

**Comment by cpascual** [comment]
mmm... I [clearly broke the formatters](https://travis-ci.org/taurus-org/taurus/builds/653369204) with my refactoring... I'll investigate and fix

**Comment by cpascual** [comment]
All clear now. Merging
